service/amount_out.go

```diff
@@ -4,130 +4,114 @@ import (
 	"context"
 	"fmt"
 
-	"github.com/google/uuid"
 	"github.com/orbs-network/order-book/models"
 	"github.com/orbs-network/order-book/utils/logger"
 	"github.com/orbs-network/order-book/utils/logger/logctx"
 	"github.com/shopspring/decimal"
 )
 
-func (s *Service) GetQuote(ctx context.Context, symbol models.Symbol, side models.Side, amountIn decimal.Decimal) (models.AmountOut, error) {
+func (s *Service) GetQuote(ctx context.Context, symbol models.Symbol, side models.Side, inAmount decimal.Decimal) (models.QuoteRes, error) {
 
-	// make sure amountIn is positivr
-	if !amountIn.IsPositive() {
-		return models.AmountOut{}, models.ErrInAmount
+	// make sure inAmount is positivr
+	if !inAmount.IsPositive() {
+		return models.QuoteRes{}, models.ErrInAmount
 	}
 	var it models.OrderIter
-	var res models.AmountOut
+	var res models.QuoteRes
 	var err error
 	if side == models.BUY {
 		it = s.orderBookStore.GetMinAsk(ctx, symbol)
-		res, err = getAmountOutInAToken(ctx, it, amountIn)
+		if it == nil {
+			logctx.Error(ctx, "GetMinAsk failed")
+			return models.QuoteRes{}, models.ErrIterFail
+		}
+		res, err = getOutAmountInAToken(ctx, it, inAmount)
 
 	} else { // SELL
 		it = s.orderBookStore.GetMaxBid(ctx, symbol)
-		res, err = getAmountOutInBToken(ctx, it, amountIn)
+		if it == nil {
+			logctx.Error(ctx, "GetMaxBid failed")
+			return models.QuoteRes{}, models.ErrIterFail
+		}
+		res, err = getOutAmountInBToken(ctx, it, inAmount)
 	}
 	if err != nil {
-		logctx.Error(ctx, "getAmountOutIn failed", logger.Error(err))
-		return models.AmountOut{}, err
+		logctx.Error(ctx, "getQuoteResIn failed", logger.Error(err))
+		return models.QuoteRes{}, err
 	}
 
 	return res, nil
 }
 
-// orderID->amount bought or sold in A token always
-func (s *Service) GetAmountOut(ctx context.Context, swapId uuid.UUID, symbol models.Symbol, side models.Side, amountIn decimal.Decimal) (models.AmountOut, error) {
-
-	var it models.OrderIter
-	var res models.AmountOut
-	var err error
-	if side == models.BUY {
-		it = s.orderBookStore.GetMinAsk(ctx, symbol)
-		res, err = getAmountOutInAToken(ctx, it, amountIn)
-
-	} else { // SELL
-		it = s.orderBookStore.GetMaxBid(ctx, symbol)
-		res, err = getAmountOutInBToken(ctx, it, amountIn)
-	}
-	if err != nil {
-		logctx.Error(ctx, "getAmountOutIn failed", logger.Error(err))
-		return models.AmountOut{}, err
-	}
-	err = s.orderBookStore.StoreSwap(ctx, swapId, res.OrderFrags)
-	if err != nil {
-		logctx.Error(ctx, "StoreSwap failed", logger.Error(err))
-		return models.AmountOut{}, err
-	}
-	return res, nil
-}
-
 // PAIR/SYMBOL A-B (ETH-USDC)
 // amount in B token (USD)
 // amount out A token (ETH)
-func getAmountOutInAToken(ctx context.Context, it models.OrderIter, amountInB decimal.Decimal) (models.AmountOut, error) {
-	amountOutA := decimal.NewFromInt(0)
+func getOutAmountInAToken(ctx context.Context, it models.OrderIter, inAmountB decimal.Decimal) (models.QuoteRes, error) {
```

Comment:
I believe the logic here looks fine, but could really use some decent unit tests to confirm and prevent future regressions! 👍

Reply:
we have tests
func TestService_getOutAmountInAToken(t *testing.T) 
/order-book/service/amount_out_test.go